src/main/java/relations/r1cs/FileToR1CS.java

```diff
@@ -0,0 +1,478 @@
+//
+//                Tuple2<LinearCombination<FieldT>, BufferedReader> resC = makeRowAt(currRow, brC);
+//                final LinearCombination<FieldT> C = resC._1();
+//                brC = resC._2();
+//
+//                constraints.add(new R1CSConstraint<>(A, B, C));
+//            }
+//
+//            brA.close();
+//            brB.close();
+//            brC.close();
+//
+//        } catch (Exception e){
+//            System.err.println("Error: " + e.getMessage());
+//        }
+//
+//        linearCombinationA.count();
+//        linearCombinationB.count();
+//        linearCombinationC.count();
+//
+//        final R1CSConstraintsRDD<FieldT> constraints = new R1CSConstraintsRDD<>(
+//                linearCombinationA,
+//                linearCombinationB,
+//                linearCombinationC,
+//                numConstraints);
+//
+//        return new R1CSRelationRDD<>(constraints, numInputs, numAuxiliary);
+//    }
+
+    private static <FieldT extends AbstractFieldElementExpanded<FieldT>>
+        LinearCombination<FieldT>
+    JSONObjectToCombination (final JSONObject matrixRow) {
+
+        final LinearCombination<FieldT> L = new LinearCombination<>();
+
+        Iterator<String> keys = matrixRow.keySet().iterator();
+        while (keys.hasNext()) {
+            String key = keys.next();
+            BN254aFields.BN254aFr value;
+            try{
+                value = new BN254aFields.BN254aFr((String) matrixRow.get(key));
+            } catch (ClassCastException e){
+                // Handle case when key-value pairs are String: Long
+                value = new BN254aFields.BN254aFr(Long.toString((long) matrixRow.get(key)));
+            }
+            L.add(new LinearTerm<>(Long.parseLong(key), (FieldT) value));
+        }
+        return L;
+    }
+
+    private static <FieldT extends AbstractFieldElementExpanded<FieldT>>
+        Tuple2<LinearCombination<FieldT>, BufferedReader>
+    makeRowAt (int index, BufferedReader reader) {
+        // Assumes input to be ordered by row and that the last line is blank.
+        final LinearCombination<FieldT> L = new LinearCombination<>();
+
+        try {
+            String nextLine;
+            reader.mark(1);  // TODO - this is very strange.
+            while ((nextLine = reader.readLine()) != null) {
+                String[] tokens = nextLine.split(" ");
+
+                int col = Integer.parseInt(tokens[0]);
+                int row = Integer.parseInt(tokens[1]);
+                final BN254aFields.BN254aFr value = new BN254aFields.BN254aFr(tokens[2]);
+
+                if (row == index) {
+                    L.add(new LinearTerm<>(col, (FieldT) value));
+                } else if (row > index) {
+                    reader.reset();
+                    return new Tuple2<>(L, reader);
+                }
+            }
+        } catch (Exception e){
+            System.err.println("Error: " + e.getMessage());
+        }
+        return new Tuple2<>(L, reader);
```

Comment:
The reader is being returned because it has a pointer to its current line. Unless this pointer is preserved while mutated inside the function call. Is there a name for this? I wasn't sure how to go about it.

Reply:
I'd expect the Reader to be passed by reference. This means that all changes to the underlying object (e.g. advancing the point in the file it is currently at), will be reflected in other references outside of this method.

In the debugger you could inspect the returned reader and the reference that is kept outside this function and check they are the same (e.g. they point to the same point in the file).

Reply:
Done!